#!/usr/bin/env bash

# Code to download raw data was generated by https://sra-explorer.info and further adapted
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/083/SRR13989883/SRR13989883_1.fastq.gz -o Control_Hypoxia_1_R1_001.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/083/SRR13989883/SRR13989883_2.fastq.gz -o Control_Hypoxia_1_R2_001.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/084/SRR13989884/SRR13989884_1.fastq.gz -o Control_Hypoxia_2_R1_001.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/084/SRR13989884/SRR13989884_2.fastq.gz -o Control_Hypoxia_2_R2_001.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/081/SRR13989881/SRR13989881_1.fastq.gz -o Control_1_R1_001.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/081/SRR13989881/SRR13989881_2.fastq.gz -o Control_1_R2_001.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/082/SRR13989882/SRR13989882_1.fastq.gz -o Control_2_R1_001.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR139/082/SRR13989882/SRR13989882_2.fastq.gz -o Control_2_R2_001.fastq.gz

for FILE in $(ls *.fastq.gz); do
    pigz -cd $FILE | head -50000 | pigz > reads/$FILE
    rm $FILE
done


### Prepare GTF files for Chr22 only
# This will speed up the test run

## TE GTF was downloaded from Hammell lab Dropbox
pigz -d GRCh38_Ensembl_rmsk_TE.gtf.gz

# Keep on Chr22 related entries
grep "^22" GRCh38_Ensembl_rmsk_TE.gtf > GRCh38_Ensembl_rmsk_TE_chr22.gtf  
pigz GRCh38_Ensembl_rmsk_TE_chr22.gtf

## Gene GTF was downloaded from Ensembl
wget https://ftp.ensembl.org/pub/release-111/gtf/homo_sapiens/Homo_sapiens.GRCh38.111.gtf.gz
pigz -d Homo_sapiens.GRCh38.111.gtf.gz

# Keep only Chr22 related entries
grep "^22" Homo_sapiens.GRCh38.111.gtf > Homo_sapiens.GRCh38.111_chr22gtf
pigz Homo_sapiens.GRCh38.111_chr22gtf